File: sdip08-urlshorten-java-spring/src/main/java/fru1t/gsd08urlshorten/shortenurl/ShortenUrlService.java
package fru1t.gsd08urlshorten.shortenurl;

import fru1t.gsd08urlshorten.common.Snowflake;
import fru1t.gsd08urlshorten.common.util.Base62;
import fru1t.gsd08urlshorten.shortenurl.model.ShortenUrl;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class ShortenUrlService {
    private final ShortenUrlRepository shortenUrlRepository;
    private final Snowflake snowflak = new Snowflake();
    private final Base62 base62 = new Base62();

    public String createShortenUrl(String longUrl) {
        return shortenUrlRepository.findBySourceUrl(longUrl)
                .map(ShortenUrl::getShortenUrl)
                .orElseGet(() -> {
                    long id = snowflak.nextId();
                    String shortUrl = base62.encode(id);
                    shortenUrlRepository.save(ShortenUrl.createShortenUrl(id, shortUrl, longUrl));
                    return shortUrl;
                });
    }

    public String readShortenUrl(String shortenUrl) {
        return shortenUrlRepository.findByShortenUrl(shortenUrl)
                .map(ShortenUrl::getSourceUrl)
                .orElseThrow(() -> new IllegalArgumentException("단축 URL을 찾을 수 없습니다: " + shortenUrl));
    }
}

File: sdip08-urlshorten-ktor/src/main/kotlin/urlshorten/domain/shortenurl/ShortenUrlService.kt
package urlshorten.domain.shortenurl

import urlshorten.common.Base62
import urlshorten.common.Snowflake
import urlshorten.plugins.NotFoundException

class ShortenUrlService(
    private val repository: ShortenUrlRepository
) {
    private val snowflake = Snowflake()

    fun createShortenUrl(longUrl: String): String {
        // 이미 존재하는 URL인 경우 기존 단축 URL 반환
        repository.findBySourceUrl(longUrl)?.let {
            return it.shortenUrl
        }

        // 새로운 단축 URL 생성
        val id = snowflake.nextId()
        val shortUrl = Base62.encode(id)

        val shortenUrl = ShortenUrl.create(id, shortUrl, longUrl)
        repository.save(shortenUrl)

        return shortUrl
    }

    fun readShortenUrl(shortenUrl: String): String {
        return repository.findByShortenUrl(shortenUrl)?.sourceUrl
            ?: throw NotFoundException("단축 URL을 찾을 수 없습니다: $shortenUrl")
    }
}
